aptos-move/aptos-vm/src/transaction_arg_validation.rs

```diff
@@ -83,22 +101,133 @@ pub(crate) fn validate_combine_signer_and_txn_args<S: MoveResolverExt>(
+    }
     Ok(combined_args)
 }
 
-fn is_valid_txn_arg<S: MoveResolverExt>(session: &SessionExt<S>, typ: &Type) -> bool {
+// Return whether the argument is valid/allowed and whether it needs validation.
+// Validation is only needed for String arguments at the moment and vectors of them.
+fn is_valid_txn_arg<S: MoveResolverExt>(session: &SessionExt<S>, typ: &Type) -> (bool, bool) {
     use move_deps::move_vm_types::loaded_data::runtime_types::Type::*;
     match typ {
-        Bool | U8 | U64 | U128 | Address => true,
+        Bool | U8 | U64 | U128 | Address => (true, false),
         Vector(inner) => is_valid_txn_arg(session, inner),
         Struct(idx) | StructInstantiation(idx, _) => {
             if let Some(st) = session.get_struct_type(*idx) {
                 let full_name = format!("{}::{}", st.module.short_str_lossless(), st.name);
-                ALLOWED_STRUCTS.contains(&full_name)
+                match ALLOWED_STRUCTS.get(&full_name) {
+                    None => (false, false),
+                    Some(validator) => (true, validator.is_some()),
+                }
             } else {
-                false
+                (false, false)
             }
         }
-        Signer | Reference(_) | MutableReference(_) | TyParam(_) => false,
+        Signer | Reference(_) | MutableReference(_) | TyParam(_) => (false, false),
+    }
+}
+
+// Validate arguments. Walk through the arguments and according to the signature
+// validate arguments that require so.
+// TODO: This needs a more solid story and a tighter integration with the VM.
+// Validation at the moment is only for Strings and Vector of them, so we
+// manually walk the serialized args until we find a string.
+// This is obviously brittle and something to change at some point soon.
+fn validate_args<S: MoveResolverExt>(
+    session: &SessionExt<S>,
+    idxs: &[usize],
+    args: &[Vec<u8>],
+    func: &LoadedFunctionInstantiation,
+) -> Result<(), VMStatus> {
+    let mut current_size_for_strings = 0;
+    for (idx, (ty, arg)) in func.parameters.iter().zip(args.iter()).enumerate() {
+        if !idxs.contains(&idx) {
+            continue;
+        }
+        let mut cursor = Cursor::new(&arg[..]);
+        validate_arg(session, ty, &mut cursor, &mut current_size_for_strings)?;
+    }
+    Ok(())
+}
+
+// Validate a single arg. A Cursor is used to walk the serialized arg manually and correctly.
+// Only Strings and nested vector of them are valiadated.
+fn validate_arg<S: MoveResolverExt>(
+    session: &SessionExt<S>,
+    ty: &Type,
+    cursor: &mut Cursor<&[u8]>,
+    current_size_for_strings: &mut usize,
+) -> Result<(), VMStatus> {
+    use move_deps::move_vm_types::loaded_data::runtime_types::Type::*;
+    Ok(match ty {
+        Vector(inner) => {
+            // get the vector length and iterate over each element
+            let mut len = get_len(cursor)?;
+            *current_size_for_strings += len;
```

Comment:
this is not the string size but the number of strings? is this intended?

Reply:
hmmm now you are making me think about all of this...

in general the name may have not been the best but it was intended to cover everything string related. So in a vector of vectors of strings it would not just include the string size but also the vector size, that was the intent.
But now I am realizing this is not the most accurate computation and maybe it's all good given the constant and what we do but let's think about it a bit.
If I get a vector of size 100 and 0 size string this would consider 100, which is good, but it's also not considering the space for the string length. So maybe adding 1 to the string length could be better, that would cover strings of zise 0 even when not in a vector. But the vector of vectors has to be taken into account. 
Moreover by just using the length we are not really computing the proper size because the number of bytes used for bigger numbers is more than 1. 
So on a vector of vectors, if each nested vector is of size > 2^7, then each size would take more than 1 byte and if the outer was length 100, the overall usage would be at least 200.
But I also wanted to be conservative and if the overall size is bigger than our constant 64K I wanted to stop no matter what.

Overall I think this may be just fine in the sense that we have a couple fo situations:
- the size of the payload is bigger then max size in which case we fail and never enter this code
- the payload is bogus and sizes are fake. In that case we stop under one of the 2 cases: 1- we reach the end of the arguments payload 2- we do not even look because we determined that the overall length is bigger than 64K. The latter seems just a early exit at this point.

My current feeling is that the length check could be easily moved out and rely on the end of the payload as a termination. In other words I do not see how we can create problems even with fake lengths.
Overall though I also think this is fine and I do not see how it would make a valid case invalid. That is, the early termination is just a minor optimization.

Happy to remove it. Or leave things as they are.

As for the variable name it seems ok to me but also happy to change it

Reply:
personally I'd like to have explicit limit on both number of items, and individual item limit. for example, we allow at most 64 strings/vector elements and each string can have at most 255 length or something like that

Reply:
I really do not like that - lol
why would you think that is a fine story? 
It seems to limit usage with no particular advantage. If I remember correctly we already know of input strings that are of a significant size, I have hard time thinking how to explain people the limitation.
Also would that apply to strings only? I can pass any size bytearray but only a certain size string?
Don't know, I have hard time to make sense of the limitation....
thoughts?

Reply:
I agree limiting only strings looks weird, ideally we should have a well-defined arguments limit, but probably not for this PR

Reply:
I think the length limit should be related to the allowable txn size limit. We currently support a 64K size limit. We should note this accordingly. The name ought to be current_evaluated_length or something

Reply:
note that length does not indicate the length size
perhaps we should be focused on the change to the cursor?